server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java

```diff
@@ -0,0 +1,262 @@
+
+    /**
+     * Creates a new HealthPeriodicLogger.
+     * This creates a scheduled job using the SchedulerEngine framework and runs it on the current health node.
+     *
+     * @param settings the cluster settings, used to get the interval setting.
+     * @param clusterService the cluster service, used to know when the health node changes.
+     * @param client the client used to call the Health Service.
+     * @param healthService the Health Service, where the actual Health API logic lives.
+     */
+    public HealthPeriodicLogger(Settings settings, ClusterService clusterService, Client client, HealthService healthService) {
+        this.settings = settings;
+        this.clusterService = clusterService;
+        this.client = client;
+        this.healthService = healthService;
+        this.clock = Clock.systemUTC();
+        this.pollInterval = HEALTH_PERIODIC_LOGGER_POLL_INTERVAL_SETTING.get(settings);
+    }
+
+    /**
+     * Initializer method to avoid the publication of a self reference in the constructor.
+     */
+    public void init() {
+        clusterService.addListener(this);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(HEALTH_PERIODIC_LOGGER_POLL_INTERVAL_SETTING, this::updatePollInterval);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        DiscoveryNode healthNode = HealthNode.findHealthNode(event.state());
+        if (healthNode == null) {
+            this.isHealthNode = false;
+            this.cancelJob();
+            return;
+        }
+        final boolean isCurrentlyHealthNode = healthNode.getId().equals(this.clusterService.localNode().getId());
+        if (this.isHealthNode != isCurrentlyHealthNode) {
+            this.isHealthNode = isCurrentlyHealthNode;
+            if (this.isHealthNode) {
+                // we weren't the health node, and now we are
+                maybeScheduleJob();
+            } else {
+                // we were the health node, and now we aren't
+                cancelJob();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        SchedulerEngine engine = scheduler.get();
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(HEALTH_PERIODIC_LOGGER_JOB_NAME)) {
+            if (this.isHealthNode && this.currentlyRunning.compareAndExchange(false, true) == false) {
+                this.callGetHealth();
+            }
+        }
+    }
+
+    // default visibility for testing purposes
+    void callGetHealth() {
+        this.healthService.getHealth(this.client, null, false, 0, this.resultsListener);
```

Comment:
So, I have been looking at this method, and I think it might be underselling itself. It is not just calling the health API, it also wires the listener. And then that got me thinking, if this is the place were the call and the listener meet. Maybe we should limit the use of `currentlyRunning` only here. 

What would be the benefit of this? We use this to synchronise a block of code in a way. When I read the code it helps me to have a clear overview of what this flag is synchronising, it helps me to reason.

For example, we could do something like this, I added extra comments to explain what it does:
```
void tryToLogHealth() {
     // We try here to get the flag 
     if (this.currentlyRunning.compareAndExchange(false, true) == false) {
            // We ensure that the flag will only be released once when it's time
            RunOnce release = new RunOnce(() -> currentlyRunning.set(false));
            try {
                // We can use the runAfter listener to wire the release to the current listener
                ActionListener<List<HealthIndicatorResult>> listenerWithRelease = ActionListener.runAfter(
                        resultsListener,
                        release
                );
                this.healthService.getHealth(this.client, null, false, 0, listenerWithRelease);
            } catch (Exception e) {
                logger.warn(() -> "The health periodic logger encountered an error.", e);
                // In case of an exception before the listener was wired, we can release the flag here, and we feel safe 
                // that it will not release it again because this can only be run once.
                release.run();
            }
        }
 }
```
This way, all the acquiring and release of the flag is one place and it's easy to share the release code but also to reason about it. What do you think?

Extra bonus, the result listener is focused only in processing the results and it doesn't have to bother with the flag.

Finally, I think it makes easier to test for that calls will not be run in parallel in a unit test.